sound/soc/sof/ipc4-topology.c

```diff
@@ -0,0 +1,836 @@
+// SPDX-License-Identifier: (GPL-2.0-only OR BSD-3-Clause)
+//
+// This file is provided under a dual BSD/GPLv2 license.  When using or
+// redistributing this file, you may do so under either license.
+//
+// Copyright(c) 2021 Intel Corporation. All rights reserved.
+//
+//
+
+#include <linux/bits.h>
+#include <linux/device.h>
+#include <linux/errno.h>
+#include <linux/firmware.h>
+#include <linux/workqueue.h>
+#include <sound/pcm_params.h>
+#include <sound/sof/topology.h>
+#include <uapi/sound/sof/tokens.h>
+#include "ops.h"
+#include "sof-audio.h"
+#include "sof-priv.h"
+#include "ipc4.h"
+#include "ipc4-topology.h"
+#include "topology-common.h"
+
+static const struct sof_topology_token ipc4_sched_tokens[] = {
+	{SOF_TKN_SCHED_LP_MODE, SND_SOC_TPLG_TUPLE_TYPE_WORD, get_token_u32,
+		offsetof(struct sof_ipc4_pipeline, lp_mode), 0}
+};
+
+/**
+ * sof_ipc4_comp_alloc - allocate and initialize buffer for a new component
+ * @swidget: pointer to struct snd_sof_widget containing extended data
+ * @ipc_size: IPC payload size that will be updated depending on valid
+ *  extended data.
+ * @index: ID of the pipeline the component belongs to
+ *
+ * Return: The pointer to the new allocated component, NULL if failed.
+ */
+static struct sof_ipc_comp *sof_ipc4_comp_alloc(struct snd_sof_widget *swidget,
```

Comment:
why cant we reuse cof_comp_alloc() here?

Reply:
IPC4 doesn't need to set hdr.cmd.  Since IPC4 doesn't use hdr we can reuse it